lib/spack/spack/bootstrap.py

```diff
@@ -28,6 +31,208 @@
 import spack.util.path
 from spack.util.environment import EnvironmentModifications
 
+#: Map a bootstrapper type to the corresponding class
+_bootstrap_methods = {}
+
+
+def _bootstrapper(type):
+    """Decorator to register classes implementing bootstrapping
+    methods.
+
+    Args:
+        type (str): string identifying the class
+    """
+    def _register(cls):
+        _bootstrap_methods[type] = cls
+        return cls
+    return _register
+
+
+@_bootstrapper(type='store')
+class _StoreBootstrapper(object):
+    def __init__(self, conf):
+        self.conf = conf
+
+    @staticmethod
+    def try_import(module, abstract_spec):
+        """Return True if the module can be imported from an already
+        installed spec, False otherwise.
+
+        Args:
+            module: Python module to be imported
+            abstract_spec: abstract spec that may provide the module
+        """
+        abstract_spec = abstract_spec or module
+
+        bincache_platform = spack.architecture.real_platform()
+        if str(bincache_platform) == 'cray':
+            bincache_platform = spack.platforms.linux.Linux()
+            with spack.architecture.use_platform(bincache_platform):
+                abstract_spec = str(spack.spec.Spec(abstract_spec))
+
+        # We have to run as part of this python interpreter
+        abstract_spec += ' ^' + spec_for_current_python()
+
+        # Check if any of the already installed specs provide the import
+        msg = "[BOOTSTRAP MODULE {0}] Try installed specs with query '{1}'"
+        tty.debug(msg.format(module, abstract_spec))
+
+        installed_specs = spack.store.db.query(abstract_spec, installed=True)
+
+        for candidate_spec in installed_specs:
+            module_path = os.path.join(
+                candidate_spec.prefix,
+                candidate_spec['python'].package.fallback_site_packages_dir
+            )
+            sys.path.append(module_path)
+
+            if _python_import(module):
+                return True
+
+            tty.warn("Spec {0} did not provide module {1}".format(
+                candidate_spec, module
+            ))
+            sys.path = sys.path[:-1]
```

Comment:
Use a try/except block so we guarantee undoing things if we fail

Reply:
Fixed in c50553a